src/Catalog/Helpers/GalleryDatabaseQueryService.cs

```diff
@@ -0,0 +1,166 @@
+                DB2CatalogCursor.ByLastEdited(since, top));
+        }
+
+        /// <summary>
+        /// Returns a <see cref="SortedList{DateTime, IList{FeedPackageDetails}}"/> from the gallery database.
+        /// </summary>
+        /// <param name="keyDateFunc">The <see cref="DateTime"/> field to sort the <see cref="FeedPackageDetails"/> on.</param>
+        private async Task<SortedList<DateTime, IList<FeedPackageDetails>>> GetPackagesInOrder(
+            Func<FeedPackageDetails, DateTime> keyDateFunc,
+            DB2CatalogCursor cursor)
+        {
+            var allPackages = await GetPackages(cursor);
+
+            return OrderPackagesByKeyDate(allPackages, keyDateFunc);
+        }
+
+        /// <summary>
+        /// Returns a <see cref="SortedList{DateTime, IList{FeedPackageDetails}}"/> of packages.
+        /// </summary>
+        /// <param name="keyDateFunc">The <see cref="DateTime"/> field to sort the <see cref="FeedPackageDetails"/> on.</param>
+        internal static SortedList<DateTime, IList<FeedPackageDetails>> OrderPackagesByKeyDate(
+            IReadOnlyCollection<FeedPackageDetails> packages,
+            Func<FeedPackageDetails, DateTime> keyDateFunc)
+        {
+            var result = new SortedList<DateTime, IList<FeedPackageDetails>>();
+
+            foreach (var package in packages)
+            {
+                var packageKeyDate = keyDateFunc(package);
+                if (!result.TryGetValue(packageKeyDate, out IList<FeedPackageDetails> packagesWithSameKeyDate))
+                {
+                    packagesWithSameKeyDate = new List<FeedPackageDetails>();
+                    result.Add(packageKeyDate, packagesWithSameKeyDate);
+                }
+
+                packagesWithSameKeyDate.Add(package);
+            }
+
+            var packagesCount = 0;
+            var filteredResult = new SortedList<DateTime, IList<FeedPackageDetails>>();
+            foreach (var keyDate in result.Keys)
+            {
+                if (result.TryGetValue(keyDate, out IList<FeedPackageDetails> packagesForKeyDate))
+                {
+                    if (packagesCount > 0 && packagesCount + packagesForKeyDate.Count > Constants.MaxPageSize)
+                    {
+                        break;
+                    }
+
+                    packagesCount += packagesForKeyDate.Count;
+                    filteredResult.Add(keyDate, packagesForKeyDate);
+                }
+            }
+
+            return filteredResult;
+        }
+
+        /// <summary>
+        /// Builds the SQL query string for db2catalog.
+        /// </summary>
+        /// <param name="cursor">The <see cref="DB2CatalogCursor"/> to be used.</param>
+        /// <returns>The SQL query string for the db2catalog job, build from the the provided <see cref="DB2CatalogCursor"/>.</returns>
+        internal static string BuildDB2CatalogSqlQuery(DB2CatalogCursor cursor)
+        {
+            return $@"SELECT TOP {cursor.Top} WITH TIES
```

Comment:
Could you paste the SQL generated by gallery for the `LastEdited`/`Created` OData queries? It's probably worth having multiple eyes on the existing SQL to made sure it's really 1:1

Reply:
The OData query is likely going to pull in a lot of unnecessary fields because the shape of the object is formed on the client-side--I don't think it's a big deal if it's not 1:1.

Reply:
True. I'm more interested in any DML that changes the number of records in the results (e.g. `WHERE`)

Reply:
The OData queries had the following format:
`https://www.nuget.org/api/v2/Packages?$filter=Created%20gt%20DateTime%272019-05-24T00:00:00%27&$orderby=Created&$top=20&$select=Id,NormalizedVersion,Created,LastEdited,Published,LicenseNames,LicenseReportUrl&semVerLevel=2.0.0`

The generated SQL query looked like this:

```
DECLARE @p__linq__5 AS SQL_VARIANT;
DECLARE @p__linq__6 AS SQL_VARIANT;
DECLARE @p__linq__4 AS SQL_VARIANT;
SET @p__linq__5 = NULL;
SET @p__linq__6 = NULL;
SET @p__linq__4 = NULL;

SELECT TOP (20) 
    [Project1].[PackageRegistrationKey] AS [PackageRegistrationKey], 
    [Project1].[C3] AS [C1], 
    [Project1].[C4] AS [C2], 
    [Project1].[Version] AS [Version], 
    [Project1].[C5] AS [C3], 
    [Project1].[C2] AS [C4], 
    [Project1].[C6] AS [C5], 
    [Project1].[C1] AS [C6], 
    [Project1].[C7] AS [C7], 
    [Project1].[C8] AS [C8], 
    [Project1].[C9] AS [C9], 
    [Project1].[LastEdited] AS [LastEdited], 
    [Project1].[C10] AS [C10], 
    [Project1].[C11] AS [C11], 
    [Project1].[C12] AS [C12], 
    [Project1].[NormalizedVersion] AS [NormalizedVersion], 
    [Project1].[C13] AS [C13], 
    [Project1].[Id] AS [Id]
    FROM ( SELECT 
        [Extent1].[PackageRegistrationKey] AS [PackageRegistrationKey], 
        [Extent1].[Created] AS [Created], 
        [Extent1].[LastEdited] AS [LastEdited], 
        [Extent1].[Version] AS [Version], 
        [Extent1].[NormalizedVersion] AS [NormalizedVersion], 
        [Extent2].[Id] AS [Id], 
        CASE WHEN ((@p__linq__5 <> 1) OR ([Extent1].[HideLicenseReport] = 1)) THEN CAST(NULL AS varchar(1)) ELSE [Extent1].[LicenseNames] END AS [C1], 
        CASE WHEN ((@p__linq__6 <> 1) OR ([Extent1].[HideLicenseReport] = 1)) THEN CAST(NULL AS varchar(1)) ELSE [Extent1].[LicenseReportUrl] END AS [C2], 
        N'40103e1f-cbe5-4f2e-bad3-4622bf14a541' AS [C3], 
        N'Version' AS [C4], 
        N'LicenseReportUrl' AS [C5], 
        N'LicenseNames' AS [C6], 
        N'Published' AS [C7], 
         CAST( CASE WHEN ([Extent1].[Listed] = 1) THEN [Extent1].[Published] ELSE @p__linq__4 END AS datetime2) AS [C8], 
        N'LastEdited' AS [C9], 
        N'Created' AS [C10], 
         CAST( [Extent1].[Created] AS datetime2) AS [C11], 
        N'NormalizedVersion' AS [C12], 
        N'Id' AS [C13]
        FROM  [dbo].[Packages] AS [Extent1]
        INNER JOIN [dbo].[PackageRegistrations] AS [Extent2] ON [Extent1].[PackageRegistrationKey] = [Extent2].[Key]
        WHERE (0 = [Extent1].[PackageStatusKey]) AND ([Extent1].[SemVerLevelKey] = 2 OR [Extent1].[SemVerLevelKey] IS NULL) AND ([Extent1].[Created] > convert(datetime2, '2018-01-01 00:00:00.0000000', 121))
    )  AS [Project1]
    ORDER BY [Project1].[Created] ASC
```

The only difference I see in the `WHERE` clause is the SemVer2 related filter, but I think we can safely ignore that.

Reply:
Awesome! Thanks for gathering this.